networking/src/NTP/Client.hs

```diff
@@ -146,14 +161,22 @@ startSend :: NtpMonad m => [SockAddr] -> NtpClient m -> m ()
 startSend addrs cli = do
     let timeout = ntpResponseTimeout (ncSettings cli)
     let poll    = ntpPollDelay (ncSettings cli)
-    logDebug "Sending requests"
-    liftIO . atomically . modifyTVarS (ncState cli) $ identity .= Just []
-    () <$ threadDelay timeout `race` forConcurrently addrs (flip doSend cli)
 
-    logDebug "Collecting responses"
-    handleCollectedResponses cli
-    liftIO . atomically . modifyTVarS (ncState cli) $ identity .= Nothing
-    liftIO $ threadDelay (poll - timeout)
+    _ <- concurrently (threadDelay poll) $ do
+        logDebug "Sending requests"
+        atomically . modifyTVarS (ncState cli) $ identity .= Just []
+        let sendRequests = forConcurrently addrs (flip doSend cli)
+        -- here we do only "send" part, so need to wait for some time
+        -- or till receiving all responses
+        let waitTimeout =
+                void $ race
+                    (threadDelay timeout)
+                    (atomically $ check =<< allResponsesGathered cli)
+        withAsync sendRequests $ \_ -> waitTimeout
+
+        logDebug "Collecting responses"
+        handleCollectedResponses cli
+        atomically . modifyTVarS (ncState cli) $ identity .= Nothing
```

Comment:
This code will likely never be executed because this thread can be cancelled as soon as `handleCollectedResponses` runs.